File: support/android/apk/servoapp/src/main/java/org/servo/servoshell/SettingsFragment.java
package org.servo.servoshell;

import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import com.google.android.material.switchmaterial.SwitchMaterial;

public class SettingsFragment extends Fragment {
    
    private SharedPreferences mPreferences;
    private SwitchMaterial mExperimentalSwitch;
    private SwitchMaterial mAnimatingSwitch;
    
    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mPreferences = PreferenceManager.getDefaultSharedPreferences(requireContext());
    }
    
    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        // Inflate the new Material3 layout
        return inflater.inflate(R.layout.fragment_settings, container, false);
    }
    
    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        
        // Find the switches and containers
        mExperimentalSwitch = view.findViewById(R.id.experimental_switch);
        mAnimatingSwitch = view.findViewById(R.id.animating_switch);
        View experimentalContainer = view.findViewById(R.id.experimental_container);
        View animatingContainer = view.findViewById(R.id.animating_container);
        
        // Load current preference values
        loadPreferences();
        
        // Set up click listeners for the entire row
        experimentalContainer.setOnClickListener(v -> {
            boolean newValue = !mExperimentalSwitch.isChecked();
            mExperimentalSwitch.setChecked(newValue);
            savePreference("experimental", newValue);
        });
        
        animatingContainer.setOnClickListener(v -> {
            boolean newValue = !mAnimatingSwitch.isChecked();
            mAnimatingSwitch.setChecked(newValue);
            savePreference("animating_indicator", newValue);
        });
        
        // Switches need extra listeners, apparently
        mExperimentalSwitch.setOnCheckedChangeListener((buttonView, isChecked) -> {
            // Only save if the change was initiated by user interaction
            if (buttonView.isPressed()) {
                savePreference("experimental", isChecked);
            }
        });
        
        mAnimatingSwitch.setOnCheckedChangeListener((buttonView, isChecked) -> {
            if (buttonView.isPressed()) {
                savePreference("animating_indicator", isChecked);
            }
        });
    }
    

    private void loadPreferences() {
        boolean experimental = mPreferences.getBoolean("experimental", false);
        boolean animatingIndicator = mPreferences.getBoolean("animating_indicator", false);
        
        // Update switch states
        mExperimentalSwitch.setChecked(experimental);
        mAnimatingSwitch.setChecked(animatingIndicator);
    }
    
    private void savePreference(String key, boolean value) {
        mPreferences.edit().putBoolean(key, value).apply();
    }
}

File: support/android/apk/servoapp/src/main/java/org/servo/servoshell/SettingsFragment.kt
package org.servo.servoshell

import android.content.SharedPreferences
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.compose.foundation.layout.Column
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ListItem
import androidx.compose.material3.Switch
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBar
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.platform.ComposeView
import androidx.compose.ui.res.stringResource
import androidx.core.content.edit
import androidx.fragment.app.Fragment
import androidx.preference.PreferenceManager

class SettingsFragment : Fragment() {
    private lateinit var preferences: SharedPreferences

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        preferences = PreferenceManager.getDefaultSharedPreferences(requireContext())
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View = inflater.inflate(R.layout.fragment_settings, container, false)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        view.findViewById<ComposeView>(R.id.toolbar).setContent {
            @OptIn(ExperimentalMaterial3Api::class)
            TopAppBar(
                title = { Text(stringResource(R.string.settings_title)) },
            )
        }
        view.findViewById<ComposeView>(R.id.body).setContent {
            Column {
                SettingsItem(
                    title = stringResource(R.string.settings_experimental_title),
                    summary = stringResource(R.string.settings_experimental_summary),
                    preferenceKey = "experimental",
                )
                SettingsItem(
                    title = stringResource(R.string.settings_animating_title),
                    summary = stringResource(R.string.settings_animating_summary),
                    preferenceKey = "animating_indicator",
                )
            }
        }
    }

    @Composable
    private fun SettingsItem(title: String, summary: String, preferenceKey: String) {
        ListItem(
            headlineContent = {
                Text(title)
            },
            supportingContent = {
                Text(summary)
            },
            trailingContent = {
                var checked by remember { mutableStateOf(preferences.getBoolean(preferenceKey, false)) }
                Switch(
                    checked = checked,
                    onCheckedChange = {
                        checked = it
                        preferences.edit { putBoolean(preferenceKey, it) }
                    },
                )
            },
        )
    }
}
